WalletWasabi/WabiSabi/Client/CredentialDependencies/DependencyGraph.cs

```diff
@@ -0,0 +1,350 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace WalletWasabi.WabiSabi.Client.CredentialDependencies
+{
+	public class DependencyGraph
+	{
+		public const int K = ProtocolConstants.CredentialNumber;
+
+		private DependencyGraph(IEnumerable<IEnumerable<long>> initialValues)
+		{
+			var initialValuesImmutable = initialValues.Select(x => x.ToImmutableArray()).ToImmutableArray();
+			if (initialValuesImmutable.Any(x => x.Length != (int)CredentialType.NumTypes))
+			{
+				throw new ArgumentException($"Number of credential values must be {CredentialType.NumTypes}");
+			}
+			if (initialValuesImmutable.Any(x => x.Where(y => y != 0).Select(y => y.CompareTo(0)).Distinct().Count() != 1))
+			{
+				throw new ArgumentException($"Credential values of a node must all have the same sign.");
+			}
+
+			for (CredentialType i = 0; i < CredentialType.NumTypes; i++)
+			{
+				if (initialValuesImmutable.Sum(x => x[(int)i]) < 0)
+				{
+					throw new ArgumentException("Overall balance must not be negative");
+				}
+			}
+
+			// per node entries created in AddNode, querying nodes not in the
+			// graph should result in key errors.
+
+			Successors = Enumerable.Repeat(0, (int)CredentialType.NumTypes).Select(_ => new Dictionary<int, HashSet<CredentialDependency>>()).ToImmutableArray();
```

Comment:
I think for private properties it makes no sense to have Immutable. For public it might be but that is not a coding convention in Wasabi IMO so before start using these you have to convince the team. Now it just adds noise, and it will cause confusion.

Reply:
An immutable array is just a subset of the API offered by a mutable array, so I don't see how this can truly add confusion, maybe it's unfamiliar but I think it can only remove confusion: what it says is "this is an array, and I promise it's not going to change, so you don't have to think about what happens if it does". This is important to me from a maintenance point of view, the algorithm in this class is complicated enough and there's plenty of mutable state to reason about as edges are added to the graph, if I have to revisit this code I want these reassurances where I can have them.

Reply:
FWIW, see this commit: https://github.com/nothingmuch/WalletWasabi/commit/991bd3a868e7a5c93d9150c53a72c93112466639

ostensibly it's a simple cleanup, but this change that should be trivial actually causes infinite loops, I ran out of patience trying to debug it but I think what's going on is that the `EdgeBalance` field of the tuple is not actually getting updated (I don't have any other explanation), but it doesn't go away even after assigning the entire tuple into the mutable array entry.

Immutable types might not be the most elegant or concise, but it's very hard to get into these weird time wasting situations when working with them, especially in concurrent settings.